### PR TITLE
Scalar qd8-f16-qb4w meta-kernel + tests

### PR DESCRIPTION
**Pull request by GregoryComer**
Add meta kernel for qd8-f16-qb4w. Generate kernels for 1x2, 2x2, 2x4, 2x8, and 4x4. Note that the meta kernel is currently located in a dedicated file, not in the main ds8-gemm/scalar.in.c. Manual test coverage. Automated test coverage and benchmarking will come in a follow-up PR (gemm test/benchmark infrastructure needs to be updated in a number of places to handle blockwise, so I'm treating that a separate task).

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -2786,6 +2792,12 @@ IF(XNNPACK_BUILD_TESTS)
   TARGET_LINK_LIBRARIES(qd8-f32-qc4w-gemm-minmax-test PRIVATE gemm-microkernel-tester hardware-config logging microkernels-all microparams-init)
   ADD_TEST(NAME qd8-f32-qc4w-gemm-minmax-test COMMAND qd8-f32-qc4w-gemm-minmax-test)
 
+  ADD_EXECUTABLE(qd8-f32-qb4w-gemm-minmax-test test/qd8-f32-qb4w-gemm-minmax.cc)
```
Comment (GregoryComer) [inline CMakeLists.txt:2795]: This is a leftover that I missed when cleaning up. I need to remove it from the PR.

File: src/qs8-gemm/scalar-qd8-f16-qb4w.c.in [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+    void* __restrict__ c,
+    size_t cm_stride,
+    size_t cn_stride,
+    const union ${PARAMS_UNION} params[XNN_MIN_ELEMENTS(1)],
+    const struct ${SCALING_PARAMS} quantization_params[XNN_MIN_ELEMENTS(${MR})])
+{
+  assert(mr != 0);
+  assert(mr <= ${MR});
+  assert(nc != 0);
+  assert(kc != 0);
```
Comment (digantdesai) [inline src/qs8-gemm/scalar-qd8-f16-qb4w.c.in:41]: perhaps we should add some bl checks to catch dumb bugs in tests like,

```
assert (bl != 0);
assert (bl <= kc);
```

File: src/qs8-gemm/scalar-qd8-f16-qb4w.c.in [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+    uint16_t* c${M} = (uint16_t*) ((uintptr_t) c + cm_stride * ${M});
+
+  kc = round_up_po2(kc, 2);
+  do {
+    $for N in range(NR):
+        const float vksum${N} = unaligned_indexed_load_f32(w, ${N});
+    w = (const float*) w + ${NR};
+
+    $for M in range(MR):
+        const int32_t vinput_zero_point${M} = quantization_params[${M}].zero_point;
```
Comment (digantdesai) [inline src/qs8-gemm/scalar-qd8-f16-qb4w.c.in:54]: Nit:

```suggestion
        const float vinput_zero_point${M} = (float) quantization_params[${M}].zero_point;
```

This is somewhat similar to what I have in my neondot and neoni8mm metakernels. Also we can do something around just converting the zero_point dtype to float but someone needs to cast it so this is cast isn't too bad.

File: src/qs8-gemm/scalar-qd8-f16-qb4w.c.in [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+
+    $for M in range(MR):
+        $for N in range(NR):
+            float vout${M}x${N} = (float) vinput_zero_point${M} * vksum${N};
+
+    size_t n_blocks = kc / bl;
```
Comment (digantdesai) [inline src/qs8-gemm/scalar-qd8-f16-qb4w.c.in:60]: what about `kc % bl != 0` - I don't believe if we can reach here with this before hitting some asserts in other parts but just makes me wonder if we should add any assert.

But conservatively, if we can't handle kc % bl != 0 in this kernel then we should assert, is what I was thinking. WDYT?

File: src/qs8-gemm/scalar-qd8-f16-qb4w.c.in [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+
+    for (size_t nb=0; nb<n_blocks; ++nb){
+      $for M in range(MR):
+        $for N in range(NR):
+            int32_t vacc${M}x${N} = 0;
+
+      for (size_t k=bl; k >= 2 * sizeof(uint8_t); k -= 2 * sizeof(uint8_t)) {
```
Comment (digantdesai) [inline src/qs8-gemm/scalar-qd8-f16-qb4w.c.in:67]: Nit:
```suggestion
      for (size_t k = bl; k >= 2 * sizeof(int8_t); k -= 2 * sizeof(int8_t)) {
```

File: src/xnnpack/microfnptr.h [hunk 6]
```diff
@@ -326,7 +326,7 @@ typedef void (*xnn_qd8_f32_qb4w_gemm_ukernel_fn)(
     const int8_t* a,
     size_t a_stride,
     const void* w,
-    float* c,
+    void* c,
```
Comment (digantdesai) [inline src/xnnpack/microfnptr.h:329]: is this intentional for fp16?

File: test/qd8-f16-qb4w-gemm-minmax.cc [hunk 7]
```diff
@@ -0,0 +1,244 @@
+// Copyright (c) Facebook, Inc. and its affiliates.
```
Comment (digantdesai) [inline test/qd8-f16-qb4w-gemm-minmax.cc:1]: I am assuming you are still working on expanding these?

**Review by digantdesai — APPROVED** [review]
Thanks. This is a great first pass. Let's discuss finer next steps in context of this particular PR in an IM.

File: test/gemm-microkernel-tester.cc [hunk 8]
```diff
@@ -1345,16 +1345,20 @@ void GemmMicrokernelTester::Test(
 
   const size_t planes = 2;  // 4 bit is 2 planes - low nibbles and high nibbles
   const size_t k2 =  round_up_po2(k(), 2);  // tester assumes byte aligned rows
+
+  // TODO - Add bl to gen script.
+  const size_t bl = k2;
```
Comment (digantdesai) [inline test/gemm-microkernel-tester.cc:1350]: this is a can of worms though. We need to discuss this in terms of what constraints are OK for this from API point of view. IMO we should pad `k` to make sure k2 % bl == 0. and some constraints what we already have on bl mainly min, max, and other rounding ones.

**Pull request merged**